### PR TITLE
Fix S2094 FP: Implicit parameterless constructor widens the scope of the base class constructor

### LINKED ISSUE #7591: Fix S2094 FP: Implicit parameterless constructor widens the scope of the base class constructor
### Description

S2094 regarding an empty class can be raised incorrectly in situations where an implicit parameterless constructor widens the scope modifier of the base class constructor. E.g. where an inherited class publishes a previously protected parameterless constructor.

### Repro steps

```cs
public class Foo
{
  protected Foo() { }
}

public class Bar : Foo
{
}
```

### Expected behavior

S2094 is not raised where the implicit parameterless constructor widens the scope modifier of the base class parameterless constructor.

### Actual behavior

S2094 is raised.

### Known workarounds

Explicitly implement the constructor. E.g.

```cs
public class Foo
{
  protected Foo() { }
}

public class Bar : Foo
{
  public Bar() : base() { }
}
```

However, this (correctly!) raises S3253 regarding having defined a redundant constructor.
In effect, the only way to fix this is to suppress S2094.

### Related information

* C#/VB.NET Plugins version : C# 9.4.0.72892 (SonarLint  7.0.0.74072)
* Visual Studio version: 17.6.5
* MSBuild / dotnet version: dotnet 7.0.306
* Operating System: Windows 11

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixes #7591
Follow-up to #9081

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Rules/ClassShouldNotBeEmptyBase.cs [hunk 1]
```diff
@@ -79,5 +79,8 @@ public abstract class ClassShouldNotBeEmptyBase<TSyntaxKind, TDeclarationSyntax>
 
     private static bool HasNonPublicDefaultConstructor(TDeclarationSyntax declaration, SemanticModel model) =>
         model.GetDeclaredSymbol(declaration) is INamedTypeSymbol classSymbol
-        && classSymbol.BaseType.Constructors.FirstOrDefault(x => x.Parameters.Length == 0) is { DeclaredAccessibility: not Accessibility.Public };
+        && classSymbol.BaseType.Constructors.FirstOrDefault(x => x.Parameters.Length == 0) is { } constructor
+        && (constructor.GetEffectiveAccessibility() < classSymbol.DeclaredAccessibility
+            // GetEffectiveAccessibility is not handling Protected
+            || (constructor.DeclaredAccessibility == Accessibility.Protected && classSymbol.DeclaredAccessibility > Accessibility.Protected));
```
Comment (CristianAmbrosini) [inline analyzers/src/SonarAnalyzer.Common/Rules/ClassShouldNotBeEmptyBase.cs:85]: Why is this one not in the `GetEffectiveAccessibility` method?

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs [hunk 2]
```diff
@@ -141,29 +141,58 @@ class Message { }                            // Compliant, ignored because of th
...
+    {
+        protected InternalClassWithProtectedConstructor() { }
+    }
+
+    public class ConstructorIsPublicAlready1 : PublicClassWithPublicConstructor { }             // Noncompliant
+    public class WidensConstructorAccessibility1 : PublicClassWithInternalConstructor { }       // Compliant
+    public class WidensConstructorAccessibility2 : PublicClassWithProtectedConstructor { }      // Compliant
 
-    class BaseWithDefaultConstructor { }                                            // Noncompliant
+    internal class ConstructorIsPublicAlready2 : PublicClassWithPublicConstructor { }           // Noncompliant
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs:182]:
```suggestion
    internal class ConstructorIsInternalAlready2 : PublicClassWithPublicConstructor { }           // Noncompliant
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs [hunk 3]
```diff
@@ -141,29 +141,58 @@ class Message { }                            // Compliant, ignored because of th
...
+    internal class ConstructorIsInternalAlready1 : PublicClassWithInternalConstructor { }       // Noncompliant
+    internal class WidensConstructorAccessibility3 : PublicClassWithProtectedConstructor { }    // Compliant
+    internal class ConstructorIsInternalAlready2 : InternalClassWithPublicConstructor { }       // Noncompliant
+    internal class ConstructorIsInternalAlready3 : InternalClassWithInternalConstructor { }     // Noncompliant
+    internal class WidensConstructorAccessibility4 : InternalClassWithProtectedConstructor { }  // Compliant
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs:187]: What about adding a few more TC with `protected class` and various combinations?

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs [hunk 4]
```diff
@@ -141,29 +141,58 @@ class Message { }                            // Compliant, ignored because of th
...
 
-    class ConstructorIsPublicAlready2 : BaseWithDefaultConstructor { }              // Noncompliant
+    class ClassWithNestedClasses
+    {
+        private class PrivateClass1 : PublicClassWithPublicConstructor { }                      // Noncompliant
+        private class PrivateClass2 : PublicClassWithInternalConstructor { }                    // Noncompliant
+        private class PrivateClass3 : PublicClassWithProtectedConstructor { }                   // Noncompliant
+        private class PrivateClass4 : InternalClassWithPublicConstructor { }                    // Noncompliant
+        private class PrivateClass5 : InternalClassWithInternalConstructor { }                  // Noncompliant
+        private class PrivateClass6 : InternalClassWithProtectedConstructor { }                 // Noncompliant
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs:196]: `private class PrivateClass6 : ProtectedClassWith###Constructor { }`
and 
`private class PrivateClass6 : PrivateClassWith###Constructor { }`

File: analyzers/src/SonarAnalyzer.Common/Rules/ClassShouldNotBeEmptyBase.cs [hunk 5]
```diff
@@ -79,5 +79,8 @@ public abstract class ClassShouldNotBeEmptyBase<TSyntaxKind, TDeclarationSyntax>
 
     private static bool HasNonPublicDefaultConstructor(TDeclarationSyntax declaration, SemanticModel model) =>
         model.GetDeclaredSymbol(declaration) is INamedTypeSymbol classSymbol
-        && classSymbol.BaseType.Constructors.FirstOrDefault(x => x.Parameters.Length == 0) is { DeclaredAccessibility: not Accessibility.Public };
+        && classSymbol.BaseType.Constructors.FirstOrDefault(x => x.Parameters.Length == 0) is { } constructor
+        && (constructor.GetEffectiveAccessibility() < classSymbol.DeclaredAccessibility
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/Rules/ClassShouldNotBeEmptyBase.cs:83]: It's an enum, so just numbers :)

File: analyzers/src/SonarAnalyzer.Common/Rules/ClassShouldNotBeEmptyBase.cs (hunk 1 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/Rules/ClassShouldNotBeEmptyBase.cs:85]: But our implementation of `GetEffectiveAccessibility`. The reason that `protected` is ignored is most likely, that `protected` can be considered more accessible than `internal`, depending on your point of view:
```cs
public virtual class
{
    internal int NotAccessibleFromOtherAssemblies { get; }
    protected int AccessibleFromOtherAssembliesWhenInheritingTheClass { get; }
}
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs (hunk 2 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs:182]: But it's public 🤔

Comment (CristianAmbrosini) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs:159]: Yep, don't know what I saw yesterday

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs [hunk 6]
```diff
@@ -141,29 +141,55 @@ class Message { }                            // Compliant, ignored because of th
...
+        protected class ConstructorIsProtectedAlready2 : ProtectedClassWithPublicConstructor { }        // Noncompliant
+        protected class WidensConstructorAccessibility6 : ProtectedClassWithInternalConstructor { }     // Compliant
+        protected class ConstructorIsProtectedAlready3 : ProtectedClassWithProtectedConstructor { }     // Noncompliant
+
+        private class PrivateClass1 : PublicClassWithPublicConstructor { }                              // Noncompliant
+        private class PrivateClass2 : PublicClassWithInternalConstructor { }                            // Noncompliant
+        private class PrivateClass3 : PublicClassWithProtectedConstructor { }                           // Noncompliant
+        private class PrivateClass4 : InternalClassWithPublicConstructor { }                            // Noncompliant
+        private class PrivateClass5 : InternalClassWithInternalConstructor { }                          // Noncompliant
+        private class PrivateClass6 : InternalClassWithProtectedConstructor { }                         // Noncompliant
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeEmpty.cs:187]: Do you mean something like this?
```cs
public class PublicClassWithPrivateConstructor { private PublicClassWithPrivateConstructor() { } }
class DoesNotCompile : PublicClassWithPrivateConstructor { }
```
That's not possible.

**Pull request merged**